### PR TITLE
breaking: move offline queue and offline store logic out of apollo link

### PR DESCRIPTION
**Pull request by darahayes**
### Description

TL:DR all network status checks, all queueing, scheduling, persistence and replaying of mutations no longer happens inside the Apollo Link chain.

Impact:

* All offline behaviour now happens only in `client.offlineMutation`.
* `client.mutate` will not perform any offline related behaviours anymore. It's just plain `client.mutate` but with some basic conflict detection still happening in there.
* All offline queueing, scheduling, replaying, persistence, etc happens outside of the apollo link chain. Instead we use a much more generic queueing mechanism that opens the door to **great flexibility**. This refactor will allow us to support more GraphQL clients and even non GraphQL clients in the near future.

Quick Summary:

* Removes `OfflineLink` and refactors the queueing logic into `OfflineQueue`.
	* `OfflineQueue` can queue any object passed into it (alongside some metadata). In this implementation it's queueing the result of `createMutationOptions`
	* `OfflineQueue` is initialised with an `execute` which fulfils items in the queue. In this implementation it's essentially `apolloClient.mutate` but it could be any function that returns a `Promise`.
	* `OfflineQueue` can have listeners attached such as `onOperationEnqueued`, `onOperationSuccess`, `onOperationFailure` which provides the owner of the queue ways to react to these situations and perform extra steps if needed. In our client implementation, we manually add optimistic responses into the apollo cache when an operation is enqueued and we can remove the optimistic response after the operation is fulfilled or had a failure.
* Because `OfflineQueue` can queue objects of any shape, `OfflineStore` now accepts a custom `Serializer` class/object that ensures the entries from the queue can be persisted. Our implementation has an Apollo specific serializer.
* Removes `BaseLink` and refactors it into something called `BaseProcessor`
* Removes `LocalDirectiveFilterLink` and the `@OnlineOnly` directive. If you want online only mutations, simply call `client.mutate`
* Removes OfflineMutationsHandler
* Removes OperationQueueEntry

##### Checklist

- [x] `npm test` passes
- [x] `npm run build` works
- [x] tests are included
- [ ] documentation is changed or added

### CODE REVIEW

**Comment by darahayes** [comment]
Still to be done:

* remove optimistic response on failures/errors
* Allow storage serialiser to be passed in
* Handle storage failures properly (remove key from array of keys)
* Handle case where storage does not exist at all.
* Differentiate between network errors
* Update functions/optimistic responses after restart
* Better dequeue function that will also remove from storage
* Make queue use generics
* Possibly implement an onOperationRequeued Listener

**Comment by wtrocki** [comment]
Need verification changes to confirm that works fine
```
TypeError: offix_offline_1.OfflineQueue is not a constructor
    at new OfflineClient (OfflineClient.js:91)
``

File: packages/offix-client/src/OfflineClient.ts [hunk 1]
```diff
@@ -47,62 +58,135 @@ export const createClient = async (userConfig: OffixClientConfig):
...
+export class OfflineClient {
 
-  public queueListeners: OfflineQueueListener[] = [];
-  public apolloClient?: ApolloOfflineClient;
-  public store: OfflineStore;
+  /**
+   * Get access to offline store that can be used to
+   * visualize  offline  operations that are currently pending
+   */
+  public get offlineStore(): OfflineStore<MutationOptions> | undefined {
```
[the remark this answers is not in the document]
Comment (darahayes) [reply packages/offix-client/src/OfflineClient.ts]: I agree. Probably not necessary!

**Comment by wtrocki** [comment]
@darahayes One thing to notice is that it will be hard to mitigate top level impact of this changes because it can be seamless to users or break their entire application.

I think that there will be couple breaking changes to document as follow up:

- Changed typescript typings for the operations etc. which might require changes in 
- mutate function is no longer supported (requires update in our workshop)
- Some updates in js-sdk will be requried to mitigate this changes and we can hide some of the changes for end users

File: packages/offix-offline/src/offline/storage/Serializer.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import { QueueEntryOperation } from "../OfflineQueue";
+
+export interface Serializer<T> {
```
[the remark this answers is not in the document]
Comment (wtrocki) [reply packages/offix-offline/src/offline/storage/Serializer.ts]: Name to generic (can clash with the other "Serializers")

File: packages/offix-client/src/OfflineClient.ts [hunk 3]
```diff
@@ -47,62 +58,135 @@ export const createClient = async (userConfig: OffixClientConfig):
...
+  public queue: OfflineQueue<MutationOptions>;
+  // listeners that can be added by the user to handle various events coming from the offline queue
+  public queueListeners: Array<OfflineQueueListener<MutationOptions>> = [];
+  // The apollo cache that holds application state
   public cache: InMemoryCache;
-  public persistor: CachePersistor<object>;
+  // wrapper around the apollo cache for persisting it across restarts
+  public persistor?: CachePersistor<object>;
+  // captures the 'base' object a mutation is performed on. Used for conflict resolution
+  public baseProcessor: BaseProcessor;
```
[the remark this answers is not in the document]
Comment (darahayes) [reply packages/offix-client/src/OfflineClient.ts]: I think the entire `baseProcessor` thing could simply become a function as opposed to some class. WDYT?

[the remark this answers is not in the document]
Comment (wtrocki) [reply packages/offix-client/src/OfflineClient.ts]: I really do not mind how we doing it. It is simply to not expose it to user as it is pretty much implementation detail of conflict that can change etc.

File: packages/offix-offline/src/offline/OfflineQueue.ts [hunk 4]
```diff
@@ -16,113 +33,176 @@ export type OperationQueueChangeHandler = (entry: OperationQueueEntry) => void;
...
+    for (const entry of this.queue) {
+      await this.forwardOperation(entry);
+    }
+  }
+
+  public async forwardOperation(entry: QueueEntry<T>) {
+    try {
+      const result = await this.execute(entry.operation);
+      this.onForwardNext(entry, result);
+      if (entry.handlers) {
```
[the remark this answers is not in the document]
Comment (darahayes) [reply packages/offix-offline/src/offline/OfflineQueue.ts]: Just on restart

File: packages/offix-client/src/apollo/LinksBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * Default HTTP Apollo Links
+ * Provides out of the box functionality for:
+ *
+ * - Offline/Online queue
+ * - Conflict resolution
+ * - Error handling
+ */
+async function createCompositeLink(config: OffixClientConfig,
```
[the remark this answers is not in the document]
Comment (darahayes) [reply packages/offix-client/src/apollo/LinksBuilder.ts]: This is nothing new, I simply moved the old file into a different location

File: packages/offix-client/src/OfflineClient.ts [hunk 6]
```diff
@@ -111,46 +195,96 @@ export class OfflineClient implements ListenerProvider {
...
+        const mutationPromise = new Promise(async (resolve, reject) => {
+          this.queue.assignHandlersToQueueEntry(queueEntry, resolve, reject);
+        });
+
+        throw new OfflineError(mutationPromise as any);
+      }
     }
   }
 
+  // TODO - does offix-cache actually need createMutationOptions?
```
[the remark this answers is not in the document]
Comment (darahayes) [reply packages/offix-client/src/OfflineClient.ts]: The `createMutationOptions` function comes from `offix-cache` right now. It isn't used anywhere else except within `OfflineClient.offlineMutate` I'm wondering is there any strong reason why `createMutationOptions` needs to be in `offix-cache`? Additionally, `createOfflineMutationOptions` in `OfflineClient` is a wrapper that adds extra things that are needed such as `conflictBase` and the `update` function from global cache update functions.

Basically the concern is that the original `createMutationOptions` in `offix-cache` kinda seems not fit for purpose to me and I also wonder why it's something we expose.

File: packages/offix-offline/src/offline/OfflineQueue.ts [hunk 7]
```diff
@@ -16,113 +33,176 @@ export type OperationQueueChangeHandler = (entry: OperationQueueEntry) => void;
...
+    // enqueue and persist
+    this.queue.push(entry);
+    // notify listeners
+    this.onOperationEnqueued(entry.operation);
+
+    if (this.store) {
+      try {
+        await this.store.saveEntry(entry.operation);
+      } catch (err) {
+        console.error(err);
```
[the remark this answers is not in the document]
Comment (darahayes) [reply packages/offix-offline/src/offline/OfflineQueue.ts]: We can use the debug package for sure

File: packages/offix-client/src/OfflineClient.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (wtrocki) [reply packages/offix-client/src/OfflineClient.ts]: Context for is that cache package is used externally by people who want to get cache updates out of the box. It has wider usage than Offix itself. However since with this change we kinda moving away from cache itself and cache updates are under control of user we might make choice to leave offix-cache unmaintained and bring some of the code inside. 

This will probably make things much cleaner, but once again this is something that we can do as next step once we go with scheduler and separate store.

File: packages/offix-client/src/OfflineClient.ts [hunk 8]
```diff
@@ -65,7 +66,7 @@ export class OfflineClient {
   // the network status interface that determines online/offline state
   public networkStatus: NetworkStatus;
   // the offline storage interface that persists offline data across restarts
-  public store?: OfflineStore<MutationOptions>;
+  public offlineStore?: OfflineStore<MutationOptions>;
```
Comment (wtrocki) [inline packages/offix-client/src/OfflineClient.ts]: We need to mark this as breaking change

**Comment by wtrocki** [comment]
One thought about storage is that if there are apps now with offline data they will need to change storage after update. This will be uber breaking hange to add yo release notes

File: packages/offix-client/src/OfflineClient.ts (hunk 8 above)
Comment (darahayes) [reply packages/offix-client/src/OfflineClient.ts]: It is marked as breaking :)

File: packages/offix-client/src/OfflineClient.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (darahayes) [reply packages/offix-client/src/OfflineClient.ts]: Just an update on this. I've refactored the entire baseProcessor logic to be purely functional. Now this is not exposed in any way.

**Comment by darahayes** [comment]
> One thought about storage is that if there are apps now with offline data they will need to change storage after update

This is a really important point. We'll definitely need to think about this more in the future and be a lot more careful.

This PR adds the concept of an `OfflineOperationSerializer` interface which is used to serialize Queue entries to storage and to deserialize offline data back into Queue entries. I think we can use this abstraction in the future to attempt migrations if the data evolves.

The other thing is that as we move towards a scheduler implementation, then hopefully we will not care too much about what is found in the queue and in storage. More responsibility will be on the user or on third party packages that integrate with the scheduler. We may still maintain Apollo specific packages for a long time so it should definitely still be on our radar. I'm going to open an issue around this where we can collaborate on ideas and potential approaches.
